tests: Run both descriptor and legacy tests within a single test invocation

achow101 opened the pull request:
Instead of having the tests only do one type or the other, depending on command line arguments, these tests should be running with both types of wallets. Any test that calls `add_wallet_options` will be run for each wallet type that it specifies a wallet option for if none are give in the command line. If a particular wallet is specified with `--legacy-wallet` or `--descriptors`, the options are still respected.

promag reviewed (commented):
Concept ACK.

S3RK commented:
-0 on this PR.

I tend to agree with ryanofsky's arguments [here](https://github.com/bitcoin/bitcoin/pull/20267#pullrequestreview-581508843).
It's better to have explicit invocations in `test_runner.py` and faster tests, because it's easier to understand and debug.

achow101 commented:
Rebased and removed the changes to `test_runner.py`. The test runner will still run with `--legacy-wallet` and `--descriptors`. For manual invocations without either option, the updated tests will run twice, once for each type of wallet.

rajarshimaitra reviewed (commented):
tACK https://github.com/bitcoin/bitcoin/pull/20892/commits/0fb11a1ac675c0c8e86d08168e78f6c365e2f559

Verified that the tests are considering both legacy and descriptors.
`--legacy-wallet` and `--descriptor` flags are also working as expected.

I found that the following wallet tests doesn't include the new `BitcoinWalletTestFramework`, but they are similar in sense of other tests and adding the new framework runs both the tests, unlike `wallet_importmulti.py` which is a "legacy only" test. Is there any specific reason for ommiting them?

- wallet_listtransactions.py
- wallet_orphanedreward.py
- wallet_reorgrestore.py
- wallet_send.py
- wallet_signer.py
- wallet_signmessagewithaddress.py
- wallet_startup.py
- wallet_taproot.py
- wallet_transactiontime_rescan.py
- wallet_txn_clone.py
- wallet_upgradewallet.py

achow101 committed:
tests: Add --legacy-wallet and --descriptor variants to new wallet tests

Some new wallet functional tests were added, these need to
--legacy-wallet and --descriptors variants in test_runner.py

achow101 commented:
I've updated those tests to use `BitcoinWalletTestFramework` except for wallet_signer.py, wallet_taproot.py, and wallet_upgradewallet.py. The first two test features that are descriptor wallet only, and the last one is legacy wallet only.

aureleoules reviewed (commented):
tACK 0b88af678aa8bab9d6980f41cff4246ed8308604 (`test/functional/test_runner.py`).
Tested on NixOS 22.05 64 bits.  

I executed wallet-related tests with `--descriptors` only and `--legacy` only, they behaved as expected.  
Running the test with no flags run both types of test as expected.
I verified that a skipped test does not skip the other test and that a failed test does not run the other.  

It would have been nice to run tests with both flags (`--descriptors --legacy`) but it looks like it requires a lot of refactor that I believe is not worth the change.

achow101 committed:
scripted-diff: Use self.tmpdir instead of self.options.tmpdir

Instead of retrieving the actual tmpdir path from self.options.tmpdir,
retrieve it from self.tmpdir.

-BEGIN VERIFY SCRIPT-
sed -i "s/self.options.tmpdir/self.tmpdir/" $(git grep -l "self.options.tmpdir" -- ":!test/functional/test_framework/test_framework.py")
sed -i "264,271b;s/self.options.tmpdir/self.tmpdir/" test/functional/test_framework/test_framework.py
-END VERIFY SCRIPT-

achow101 committed:
tests: Always use tempfile.mkdtemp and a dedicated tmpdir variable

Instead of manually making the tmpdir depending on whether --tmpdir is
provided, we can always us tempfile.mkdtemp and provide the dir option
with self.options.tmpdir. This behaves the same as previously except
that the tmpdirs created when --tmpdir is passed will have additional
randomness on the directory names.

In order to get the correct tmpdir prefixes from the test runner, a
--tmpdir-prefix option is added.

Lastly, the tmpdir location is stored in a dedicated self.tmpdir
variable. We wll remove usage of self.options.tmpdir for that path in
later commits.

achow101 committed:
test: Remove setting self.options.tmpdir to tmpdir

MarcoFalke commented:
Not sure how useful this refactor is. As bdb will be removed, this is only a temporary patch that will be reverted later.

If this is not a refactor, it might be good to explain the goal, so that alternative solutions can be considered.

Moreover, the CI will already run a set of tests with sqlite-only and bdb-only. (Test failures due to missing build module checks seem more frequent than wallet related test failures that are not caught with a simple call to `test_runner.py` on current master)

achow101 commented:
> Not sure how useful this refactor is. As bdb will be removed, this is only a temporary patch that will be reverted later.
> 
> If this is not a refactor, it might be good to explain the goal, so that alternative solutions can be considered.
> 
> Moreover, the CI will already run a set of tests with sqlite-only and bdb-only. (Test failures due to missing build module checks seem more frequent than wallet related test failures that are not caught with a simple call to `test_runner.py` on current master)

The goal is not to refactor (that is never an outright goal for me) but rather to make it so that running a wallet test by itself (e.g. `test/functional/rpc_fundrawtransaction.py`) will run it with both `--legacy-wallet` and `--descriptors`. This is somewhat important for me because I have often forgotten to add one of those flags and accidentally thought that the tests were passing when in fact the relevant ones were not. While CI does do both and catches these failures, I think it is better to make it easier for developers to run such tests in both modes to catch errors before they are pushed to the PR.

MarcoFalke commented:
Ok, I just don't like making this a new class, which seems to conflict with the command line options.

My idea is to specify a wallet test in `set_test_params` and then provide the `--descriptors` and `--legacy` args in the parsers (a nice side effect would be to not provide them for tests that have nothing to do with the wallet).

If `set_test_params` was moved before the arg parsing, any stuff that needs args would be moved out of `set_test_params`. This way there wouldn't be several different ways to make a test "wallet-y" (class vs command line options).

Then, if no option is provided in the parser, run both legacy and descriptor wallets.

Maybe even assert in the test runner that for wallet tests exactly one option is given? This would be another nice side effect to enforce consistency at CI time.

achow101 commented:
> My idea is to specify a wallet test in `set_test_params` and then provide the `--descriptors` and `--legacy` args in the parsers (a nice side effect would be to not provide them for tests that have nothing to do with the wallet).

I'm open to reviewing alternatives.

MarcoFalke commented:
Sure, see https://github.com/bitcoin/bitcoin/pull/26480. I think (untested draft), you can drop the first two commits and replace them with a patch along the lines of:


```diff
diff --git a/test/functional/test_framework/test_framework.py b/test/functional/test_framework/test_framework.py
index d88cccecd9..61b5dfc7e9 100755
--- a/test/functional/test_framework/test_framework.py
+++ b/test/functional/test_framework/test_framework.py
@@ -100,6 +100,7 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
         self.network_thread = None
         self.rpc_timeout = 60  # Wait for up to 60 seconds for the RPC server to respond
         self.supports_cli = True
+        self._runs_wallet_type = [None]
         self.bind_to_localhost_only = True
         self.parse_args()
         self.disable_syscall_sandbox = self.options.nosandbox or self.options.valgrind
@@ -128,33 +129,48 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
 
         assert hasattr(self, "num_nodes"), "Test must set self.num_nodes in set_test_params()"
 
-        try:
-            self.setup()
-            self.run_test()
-        except JSONRPCException:
-            self.log.exception("JSONRPC error")
-            self.success = TestStatus.FAILED
-        except SkipTest as e:
-            self.log.warning("Test Skipped: %s" % e.message)
-            self.success = TestStatus.SKIPPED
-        except AssertionError:
-            self.log.exception("Assertion failed")
-            self.success = TestStatus.FAILED
-        except KeyError:
-            self.log.exception("Key error")
-            self.success = TestStatus.FAILED
-        except subprocess.CalledProcessError as e:
-            self.log.exception("Called Process failed with '{}'".format(e.output))
-            self.success = TestStatus.FAILED
-        except Exception:
-            self.log.exception("Unexpected exception caught during testing")
-            self.success = TestStatus.FAILED
-        except KeyboardInterrupt:
-            self.log.warning("Exiting after keyboard interrupt")
-            self.success = TestStatus.FAILED
-        finally:
-            exit_code = self.shutdown()
-            sys.exit(exit_code)
+        exit_codes = []  # The final exit code(s) of the test run(s). May have two values instead of one if both wallet types are run.
+        for wallet_type in self._runs_wallet_type:
+            self.options.descriptors=wallet_type
+            try:
+                self.setup()
+                self.run_test()
+            except JSONRPCException:
+                self.log.exception("JSONRPC error")
+                self.success = TestStatus.FAILED
+            except SkipTest as e:
+                self.log.warning("Test Skipped: %s" % e.message)
+                self.success = TestStatus.SKIPPED
+            except AssertionError:
+                self.log.exception("Assertion failed")
+                self.success = TestStatus.FAILED
+            except KeyError:
+                self.log.exception("Key error")
+                self.success = TestStatus.FAILED
+            except subprocess.CalledProcessError as e:
+                self.log.exception("Called Process failed with '{}'".format(e.output))
+                self.success = TestStatus.FAILED
+            except Exception:
+                self.log.exception("Unexpected exception caught during testing")
+                self.success = TestStatus.FAILED
+            except KeyboardInterrupt:
+                self.log.warning("Exiting after keyboard interrupt")
+                self.success = TestStatus.FAILED
+            finally:
+                self.shutdown()
+                exit_codes.append(self.success)
+
+        # Determine the exit code
+        # If one failed, the test failed.
+        # If both skipped, the test skipped.
+        # If one passed and the other skipped, the test passed
+        if any([ec == TestStatus.FAILED for ec in exit_codes]):
+            sys.exit(TEST_EXIT_FAILED)
+        if all([ec == TestStatus.SKIPPED for ec in exit_codes]):
+            sys.exit(TEST_EXIT_SKIPPED)
+        if any([ec == TestStatus.PASSED for ec in exit_codes]):
+            sys.exit(TEST_EXIT_PASSED)
+        assert False, "Unhandled exit scenario Legacy {}, descriptor {}".format(exit_codes[0], exit_codes[1])            
 
     def parse_args(self):
         previous_releases_path = os.getenv("PREVIOUS_RELEASES_DIR") or os.getcwd() + "/releases"
@@ -214,7 +230,9 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
         elif self.options.descriptors is None:
             # Some wallet is either required or optionally used by the test.
             # Prefer BDB unless it isn't available
-            if self.is_bdb_compiled():
+            if self.is_bdb_compiled() and self.is_sqlite_compiled():
+                self._runs_wallet_type = [False,True]
+            elif self.is_bdb_compiled():
                 self.options.descriptors = False
             elif self.is_sqlite_compiled():
                 self.options.descriptors = True

achow101 committed:
Add self.use_descriptors that is an optional bool

self.use_descriptors tracks whether the test should be using a
descriptor. However, to avoid the fact that None evaluates to False, it
is a newly introduced OptionalBool class that will assert if it is None
when used as a bool rather than returning False.

test_framework.py is updated to use self.use_descriptors throughout
rather than self.options.descriptors. The next commit will update the
rest of the tests.

achow101 committed:
scripted-diff: Replace self.options.descriptors with use_descriptors

Instead of using the options parser to hold whether we are doing
descriptor wallets, have a member in the test framework class to handle
it.

-BEGIN VERIFY SCRIPT-
sed -i -E "s/self.options.descriptors/self.use_descriptors/" $(git grep -l "self.options.descriptors" -- ":(exclude)*/test_framework.py")
-END VERIFY SCRIPT-

achow101 committed:
Remove unnecessary self.options.descriptors assignments

self.options.descriptors is no longer used by the tests directly, so we
don't need to be assigning it.

achow101 committed:
tests: Move conditional default wallet name to setup()

Determining the default wallet name based on the wallet type the test is
running with requires being in `setup()`, so move the setting of
`self.default_wallet_name` there. This requires changing a couple of
tests in the way that they can specify that the default wallet name
should be used in `self.wallet_names`.

achow101 committed:
tests: Run both descriptor and legacy wallet modes in single invocation

Be able to run tests that can run in both descriptor and legacy wallet
modes with a single invocation of the test script. In order to
facilitate this, self.options.descriptors is changed to be a list. When
neither option is provided, but both are available, it is set to [False,
True] which will run the test twice, the first in legacy wallet mode,
the second in descriptor wallet mode.

For the tests that can only have one mode, add_wallet_options will set
self.options.descriptors to either [False] or [True] which will make the
correct type run.

achow101 commented:
Rebased and reworked now that #26480 is merged. I've had to change how to decide whether to use descriptor wallets - instead of checking `self.options.descriptors`, we now check a `self.use_descriptors`.

Additionally, the default wallet name is determined conditionally based on the wallet type we are using for the test. As such ,it can't be in `__init__()`, so I moved it to `setup()`. This breaks a couple of tests that used it in their `set_test_params()`, so I've also added a special case for `True` in `self.wallet_names` to signify that `self.default_wallet_name` should be used.

maflcko reviewed (commented):
left a question

achow101 commented on test/functional/test_framework/test_framework.py:
```diff
@@ -212,18 +223,24 @@ def parse_args(self):
             # It still needs to exist and be None in order for tests to work however.
             # So set it to None to force -disablewallet, because the wallet is not needed.
             self.options.descriptors = None
+            self.use_descriptors = OptionalBool()
```
> Might be good to explain why this commit is needed. `None` will already force `-disablewallet`, and thus assert when a logic error occurs.

`self.use_descriptors` was needed because we can't be overriding `self.options.descriptors` in `main()` while also respecting its value from `parse_args()`. I also wanted to change the `self.options.descriptors` to an array of the wallet types to run with, which was incompatible with our current usage of that value.

`OptionalBool` was introduced to catch other logic errors within the tests themselves in areas that are not reliant on bitcoind running. For example, it caught the issue with the default wallet name.

> Also, it doesn't seem to be working, given that the tests fail when only sqlite is installed?

That's unrelated to this particular change. That was occurring because some tests add the wallet options since they can run with a wallet, but do not require it. For those tests, they were being run with both wallet types always, even if one type (e.g. legacy) was not available, causing the failure. I've pushed a fix for that.

achow101 commented:
Ran into some issues with `mempool_compatibility.py` so I've added some commits that remove the usage of the wallet from that.

maflcko reviewed (commented):
left a comment/nit

achow101 commented on test/functional/test_runner.py:
```diff
@@ -149,8 +149,10 @@
     'tool_wallet.py --descriptors',
```
I don't quite follow. Why would this change whether it runs sequentially or parallel?

MarcoFalke replied:
My understanding is that a missing type setting will run both types, if possible, sequentially.

The goal of test_runner is to run as much in parallel as possible. So if the test_runner setting is missing a wallet type setting in one line, it will run the test twice, sequentially.

This can be fixed by forcing a type in the test runner, thus running at most one test per line, thus all lines/tests in parallel.

achow101 replied:
Done.

There are a few tests that do `add_wallet_options` but don't really test anything with the wallet. For those tests, I've added an `any_type` to `add_wallet_options` so that those tests don't need either `--legacy-wallet` or `--descriptors` specified and won't be run twice by the test runner.

MarcoFalke commented on test/functional/mempool_compatibility.py:
```diff
@@ -37,25 +33,26 @@ def setup_network(self):
         old_node, new_node = self.nodes
-        new_wallet = MiniWallet(new_node, mode=MiniWalletMode.RAW_P2PK)
+        new_wallet = MiniWallet(new_node, mode=MiniWalletMode.RAW_P2PK_RANDOM)
+        old_wallet = MiniWallet(old_node, mode=MiniWalletMode.RAW_P2PK_RANDOM)
```
Generally MiniWallet is designed to use the latest features of Bitcoin Core, so it may not work with older versions. Also, it should be sufficient to have a single wallet instance per test.
Thus, it should be possible to remove the below `generatetodescriptor` and just replace `1` blocks with `2` in the line below. And also the "random" miniwallet mode is not needed.

MarcoFalke replied:
Done in https://github.com/bitcoin/bitcoin/pull/26640 , which should hopefully allow to drop two commits from this pull and make it simpler.

achow101 committed:
tests: Automatically run both wallet types in parallel for wallet tests

Most wallet tests need to be run with both wallet types. Instead of
requiring developers to manually specify this, the test_runner can find
the wallet tests that do not have a wallet type specified and
automatically run the test with both types.

achow101 commented:
I've added a commit which has the test runner automatically add the `--legacy-wallet` and `--descriptors` to `wallet_*` tests that don't have those options already specified. This reduces the need to remember to add those options to wallet tests.

aureleoules commented:
CI is failing on 
```
1/270 - feature_pruning.py failed, Duration: 0 s
stdout:
stderr:
usage: feature_pruning.py [options]
feature_pruning.py: error: one of the arguments --descriptors --legacy-wallet is required
```
https://cirrus-ci.com/task/5943446209298432?logs=functional_tests#L20

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -169,6 +169,7 @@ def parse_args(self):
         parser.add_argument("--cachedir", dest="cachedir", default=os.path.abspath(os.path.dirname(os.path.realpath(__file__)) + "/../../cache"),
                             help="Directory for caching pregenerated datadirs (default: %(default)s)")
         parser.add_argument("--tmpdir", dest="tmpdir", help="Root directory for datadirs")
+        parser.add_argument("--tmpdir-prefix", dest="tmpdir_prefix", help="Prefix to use for datadirs", default=TMPDIR_PREFIX)
```
nit in the first commit: This makes the `--tmpdirprefix` option pretty useless (and confusing, given that there are now 3 tmpdir options). My recommendation would be to remove `--tmpdirprefix` when adding this option.

`--tmpdirprefix` is also silently ignored when `--tmpdir` is passed, so this could even make sense as a separate bug-fix PR

achow101 replied:
I don't quite follow, `--tmpdirprefix` is for the test runner. If it can be left for a followup, I'd prefer to leave it for that.

MarcoFalke replied:
Any option for a single test will be passed up to the test runner. (Similar to how **kwargs can be used to pass through options).

If `--tmpdir` is used to set the root dir and `--tmpdir-prefix` to set the prefix folder name, and both options are passed up to test runner, then having a third and broken `--tmpdirprefix`, which only works for the test runner, seems increasingly confusing and there may even introduce new bugs.

Moreover, I wonder which later commit in this pull requires the first commit(s). Looking at 2a60a78a6483d59b3ea09377d663ca834abc7fcf, it looks like you are using the test runner to parallelize the test, which should already correctly set the unique seed id in the temp prefix folder name?

achow101 replied:
I've added a commit which remove `--tmpdirprefix`

> Moreover, I wonder which later commit in this pull requires the first commit(s).

They're necessary for 88b63a6e163045e63a7540427834a57799234407 "tests: Automatically run both wallet types in parallel for wallet tests". The single invocation needs to have different tmpdirs for each wallet option it runs with. Although `--tmpdir-prefix` might not be needed with the last commit.

achow101 replied:
Sorry, I think I gave the wrong commit. It should be a936a3f194f855e269a4dc2c25e17dbc55eb7a18 "tests: Run both descriptor and legacy wallet modes in single invocation".

The tmpdir changes are needed for this commit because the entire test is restarted for the other wallet type, so it needs to be able to make a new datadir.

achow101 committed:
tests: Remove --tmpdirprefix

--tmpdirprefix is now unnecessary and potentially confusing.

josibake commented:
Concept ACK

I think when I initially looked at this PR the `BitcoinWalletTestFramework` class was breaking `TestShell`, but looking through the PR now, it seems the approach has changed to not use a `BitcoinWalletTestFramework` class. Can you update the description detailing the approach you are now taking?

Did a brief code review and the changes good look good, will test and review once you update the description so I can determine if the code is actually doing what you want it to do :smile:

achow101 commented:
> Can you update the description detailing the approach you are now taking?

Done.

aureleoules commented:
I have the following error when running the test_runner with BDB compiled:

<details>
<summary>Details</summary>

```bash
1695642351381,"4/288 - wallet_backup.py --legacy-wallet failed, Duration: 43 s"
1695642351381,stdout:
1695642351381,2023-09-25T11:39:01.368000Z TestFramework (INFO): PRNG seed is: 8911868130033750645
1695642351381,2023-09-25T11:39:01.381000Z TestFramework (INFO): Initializing test directory /tmp/test_runner_₿_🏃_20230925_113853/wallet_backup_272_3wjtcweb
1695642351381,2023-09-25T11:39:05.172000Z TestFramework (INFO): Generating initial blockchain
1695642351381,2023-09-25T11:39:09.767000Z TestFramework (INFO): Creating transactions
1695642351381,2023-09-25T11:39:25.216000Z TestFramework (INFO): Backing up
1695642351381,2023-09-25T11:39:25.622000Z TestFramework (INFO): More transactions
1695642351381,2023-09-25T11:39:39.536000Z TestFramework (INFO): Restoring wallets on node 3 using backup files
1695642351381,2023-09-25T11:39:40.434000Z TestFramework (INFO): Restoring using dumped wallet
1695642351381,2023-09-25T11:39:43.078000Z TestFramework (ERROR): Unexpected exception caught during testing
1695642351381,Traceback (most recent call last):
1695642351381,"  File ""/tmp/bitcoin/test/functional/test_framework/test_framework.py"", line 148, in main"
1695642351381,    self.run_test()
1695642351381,"  File ""/tmp/bitcoin/test/functional/wallet_backup.py"", line 231, in run_test"
1695642351381,"    self.nodes[node_num].createwallet(wallet_name=self.default_wallet_name, descriptors=self.use_descriptors, load_on_startup=True)"
1695642351381,"  File ""/tmp/bitcoin/test/functional/test_framework/test_node.py"", line 818, in createwallet"
1695642351381,"    return self.__getattr__('createwallet')(wallet_name, disable_private_keys, blank, passphrase, avoid_reuse, descriptors, load_on_startup, external_signer)"
1695642351381,"  File ""/tmp/bitcoin/test/functional/test_framework/coverage.py"", line 50, in __call__"
1695642351381,"    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)"
1695642351381,"  File ""/tmp/bitcoin/test/functional/test_framework/authproxy.py"", line 126, in __call__"
1695642351381,"    postdata = json.dumps(self.get_request(*args, **argsn), default=serialization_fallback, ensure_ascii=self.ensure_ascii)"
1695642351381,"  File ""/tmp/bitcoin/test/functional/test_framework/authproxy.py"", line 114, in get_request"
1695642351381,"    json.dumps(args or argsn, default=serialization_fallback, ensure_ascii=self.ensure_ascii),"
1695642351381,"  File ""/usr/lib/python3.10/json/__init__.py"", line 238, in dumps"
1695642351381,    **kw).encode(obj)
1695642351381,"  File ""/usr/lib/python3.10/json/encoder.py"", line 199, in encode"
1695642351381,"    chunks = self.iterencode(o, _one_shot=True)"
1695642351381,"  File ""/usr/lib/python3.10/json/encoder.py"", line 257, in iterencode"
1695642351381,"    return _iterencode(o, 0)"
1695642351381,"  File ""/tmp/bitcoin/test/functional/test_framework/authproxy.py"", line 68, in serialization_fallback"
1695642351381,"    raise TypeError(repr(o) + "" is not JSON serializable"")"
1695642351381,TypeError: <test_framework.test_framework.OptionalBool object at 0x7fbcf6e055a0> is not JSON serializable
1695642351381,2023-09-25T11:39:43.130000Z TestFramework (INFO): Stopping nodes
1695642351381,2023-09-25T11:39:43.507000Z TestFramework (WARNING): Not cleaning up dir /tmp/test_runner_₿_🏃_20230925_113853/wallet_backup_272_3wjtcweb
1695642351381,2023-09-25T11:39:43.507000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/test_runner_₿_🏃_20230925_113853/wallet_backup_272_3wjtcweb/test_framework.log
1695642351381,2023-09-25T11:39:43.508000Z TestFramework (ERROR): 
1695642351381,2023-09-25T11:39:43.508000Z TestFramework (ERROR): Hint: Call /tmp/bitcoin/test/functional/combine_logs.py '/tmp/test_runner_₿_🏃_20230925_113853/wallet_backup_272_3wjtcweb' to consolidate all logs
1695642351381,2023-09-25T11:39:43.508000Z TestFramework (ERROR): 
1695642351381,"2023-09-25T11:39:43.508000Z TestFramework (ERROR): If this failure happened unexpectedly or intermittently, please file a bug and provide a link or upload of the combined log."
1695642351381,2023-09-25T11:39:43.508000Z TestFramework (ERROR): https://github.com/bitcoin/bitcoin/issues
1695642351381,2023-09-25T11:39:43.508000Z TestFramework (ERROR): 
```

</details>

achow101 commented:
The intention is to remove the legacy wallet soon(tm), so I guess this isn't worth merging and then reverting in a few months.
